lib/gc/Transforms/GPU/GpuUtils.h

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef GPUUTILS_H
+#define GPUUTILS_H
+
+#include "gc/Utils/Log.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/DataLayoutInterfaces.h"
+
+#include <numeric>
+
+using namespace mlir;
+
+namespace mlir::gc {
+template <typename DerivedT> struct GpuPass {
+
+  int64_t getGpuPropertyAsInt(Builder &builder, StringRef name,
+                              int64_t defaultValue) {
+    if (auto mod = static_cast<DerivedT *>(this)
+                       ->getOperation()
+                       ->template getParentOfType<ModuleOp>()) {
+      DataLayout layout(mod);
+      if (auto value = layout.getDevicePropertyValue(
+              builder.getStringAttr("GPU" /* device ID*/),
+              builder.getStringAttr(name))) {
+        if (auto attr = dyn_cast<IntegerAttr>(*value)) {
+          return attr.getInt();
+        }
+      }
+    }
+    return defaultValue;
+  }
+
+  int64_t getNumEus(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_exec_units",
+                               static_cast<DerivedT *>(this)->numEus);
+  }
+
+  int64_t getNumEusPerSlice(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_exec_units_per_slice",
+                               static_cast<DerivedT *>(this)->numEusPerSlice);
+  }
+
+  int64_t getNumThreadsPerEu(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_threads_per_eu",
+                               static_cast<DerivedT *>(this)->numThreadsPerEu);
+  }
+
+  int64_t getCacheSize(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "L1_cache_size_in_bytes",
+                               static_cast<DerivedT *>(this)->cacheSize);
+  }
+
+  int64_t getVectorWidth(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "max_vector_op_width",
+                               static_cast<DerivedT *>(this)->vectorWidth);
+  }
+
+  int64_t getWorkGroupSize(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "max_work_group_size",
+                               static_cast<DerivedT *>(this)->workGroupSize);
+  }
+};
+
+struct OpRewriter final : IRRewriter {
+  Location loc;
+
+  explicit OpRewriter(func::FuncOp &func)
+      : IRRewriter(func.getContext()), loc(func.getLoc()), func(func) {}
+
+  template <typename OpTy, typename... Args> OpTy create(Args &&...args) {
+    return RewriterBase::create<OpTy>(loc, std::forward<Args>(args)...);
+  }
+
+  arith::ConstantIndexOp getConstant(int64_t v) {
+    if (func.empty()) {
+      func.addEntryBlock();
+    } else {
+      // Try to find ConstantIndexOp with the same value in the function.
```

Comment:
I think running `--canonicalize` after the pass eliminates all duplicated constants automatically

Reply:
I know, but when working on a pass and often reviewing/debugging the produced code, for me it's a bit more convenient having a cleaner code, without duplicates. Anyway, I do not insist on having this functionality. The main purpose of OpRewriter is to have a wrapper for boilerplate code. As an example, instead of widely used `rewriter.create<arith::ConstantIndexOp>(loc, x)` we can use less verbose `rewriter.getConstant(x)`. Another example - loc. Often, this is a local variable that is passed to each call of rewriter. By overloading methods, OpRewriter adds the loc arguments to the calls - `create<arith::MulIOp>(x, y)` vs `create<arith::MulIOp>(loc, x, y)`